analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs

```diff
@@ -56,18 +60,29 @@ private static ProgramState LearnBranchingConstraint<T>(ProgramState state, IBin
         {
             return null;
         }
+    }
+
+    // We can take the 1st constraint and "testedSymbol" because they are exclusive. Symbols with NotNull constraint will be recognized as the constraining side.
+    // We only learn in the true branch because not being >, >=, <, <= than a non-empty nullable means either being null or non-null with non-matching value.
+    private static ProgramState LearnBranchingRelationalConstraint(ProgramState state, IBinaryOperationWrapper binary, bool falseBranch) =>
+        !falseBranch
+        && FirstBinaryOperandConstraint<ObjectConstraint>(state, binary) == ObjectConstraint.NotNull
+        && FirstBinaryOperandSymbolWithoutConstraint<ObjectConstraint>(state, binary) is { } testedSymbol
+        && testedSymbol.GetSymbolType().IsNullableValueType()
+            ? state.SetSymbolConstraint(testedSymbol, ObjectConstraint.NotNull)
+            : null;
 
-        ISymbol OperandSymbolWithoutConstraint(IOperation candidate) =>
-            candidate.TrackedSymbol() is { } symbol
-            && (state[symbol] is null || !state[symbol].HasConstraint<T>())
-                ? symbol
-                : null;
+    private static SymbolicConstraint FirstBinaryOperandConstraint<T>(ProgramState state, IBinaryOperationWrapper binary) where T : SymbolicConstraint =>
+        OperandConstraint<T>(state, binary.LeftOperand) ?? OperandConstraint<T>(state, binary.RightOperand);
 
-        SymbolicConstraint OperandConstraint(IOperation candidate) =>
-            state[candidate] is { } value && value.HasConstraint<T>()
-                ? value.Constraint<T>()
-                : null;
-    }
+    private static ISymbol FirstBinaryOperandSymbolWithoutConstraint<T>(ProgramState state, IBinaryOperationWrapper binary) where T : SymbolicConstraint =>
+        OperandSymbolWithoutConstraint<T>(state, binary.LeftOperand) ?? OperandSymbolWithoutConstraint<T>(state, binary.RightOperand);
+
+    private static ISymbol OperandSymbolWithoutConstraint<T>(ProgramState state, IOperation candidate) where T : SymbolicConstraint =>
+        candidate.TrackedSymbol() is { } symbol && (state[symbol] is null || !state[symbol].HasConstraint<T>()) ? symbol : null;
+
+    private static SymbolicConstraint OperandConstraint<T>(ProgramState state, IOperation candidate) where T : SymbolicConstraint =>
+        state[candidate] is { } value && value.HasConstraint<T>() ? value.Constraint<T>() : null;
```

Comment:
Is'nt this the same as
```
state[candidate]?.Constraint<T>()
```
?
In that case, we don't need to extract it as a method.

Reply:
Yes, I hadn't realized that `value.Constraint<T>()`  was returning `null`.

Simplified and inlined.